// react-spotify-web-playback/lib/components/Controls.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var jsx_runtime_1 = require("react/jsx-runtime");
var Next_1 = require("./icons/Next");
var Pause_1 = require("./icons/Pause");
var Play_1 = require("./icons/Play");
var Previous_1 = require("./icons/Previous");
var styles_1 = require("../styles");
var Wrapper = (0, styles_1.styled)('div')({}, function (_a) {
    var style = _a.style;
    return ({
        alignItems: 'center',
        display: 'flex',
        height: (0, styles_1.px)(style.h),
        justifyContent: 'center',
        '@media (max-width: 767px)': {
            padding: (0, styles_1.px)(10),
        },
        '> div': {
            minWidth: (0, styles_1.px)(style.h),
            textAlign: 'center',
        },
        button: {
            alignItems: 'center',
            color: style.c,
            display: 'inline-flex',
            fontSize: (0, styles_1.px)(16),
            height: (0, styles_1.px)(48),
            justifyContent: 'center',
            width: (0, styles_1.px)(48),
            '&.rswp__toggle': {
                fontSize: (0, styles_1.px)(28),
            },
        },
    });
}, 'ControlsRSWP');
function Controls(props) {
    var isExternalDevice = props.isExternalDevice, isPlaying = props.isPlaying, locale = props.locale, nextTracks = props.nextTracks, onClickNext = props.onClickNext, onClickPrevious = props.onClickPrevious, onClickTogglePlay = props.onClickTogglePlay, previousTracks = props.previousTracks, _a = props.styles, color = _a.color, height = _a.height;
    return ((0, jsx_runtime_1.jsxs)(Wrapper, __assign({ style: { c: color, h: height } }, { children: [(0, jsx_runtime_1.jsx)("div", { children: (!!previousTracks.length || isExternalDevice) && ((0, jsx_runtime_1.jsx)("button", __assign({ "aria-label": locale.previous, onClick: onClickPrevious, title: locale.previous, type: "button" }, { children: (0, jsx_runtime_1.jsx)(Previous_1.default, {}) }))) }), (0, jsx_runtime_1.jsx)("div", { children: (0, jsx_runtime_1.jsx)("button", __assign({ "aria-label": isPlaying ? locale.pause : locale.play, className: "rswp__toggle", onClick: onClickTogglePlay, title: isPlaying ? locale.pause : locale.play, type: "button" }, { children: isPlaying ? (0, jsx_runtime_1.jsx)(Pause_1.default, {}) : (0, jsx_runtime_1.jsx)(Play_1.default, {}) })) }), (0, jsx_runtime_1.jsx)("div", { children: (!!nextTracks.length || isExternalDevice) && ((0, jsx_runtime_1.jsx)("button", __assign({ "aria-label": locale.next, onClick: onClickNext, title: locale.next, type: "button" }, { children: (0, jsx_runtime_1.jsx)(Next_1.default, {}) }))) })] })));
}
exports.default = Controls;
