// react-spotify-web-playback/lib/components/Content.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var jsx_runtime_1 = require("react/jsx-runtime");
var styles_1 = require("../styles");
var Wrapper = (0, styles_1.styled)('div')({
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    position: 'relative',
    '> *': {
        width: '100%',
        '@media (min-width: 1024px)': {
            width: '33.3333%',
        },
    },
    '@media (min-width: 1024px)': {
        flexDirection: 'row',
    },
}, function (_a) {
    var style = _a.style;
    return ({
        minHeight: (0, styles_1.px)(style.h),
    });
}, 'ContentRSWP');
function Content(_a) {
    var children = _a.children, styles = _a.styles;
    return (0, jsx_runtime_1.jsx)(Wrapper, __assign({ style: { h: styles.height } }, { children: children }));
}
exports.default = Content;
