// react-spotify-web-playback/lib/components/ErrorMessage.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var jsx_runtime_1 = require("react/jsx-runtime");
var styles_1 = require("../styles");
var Wrapper = (0, styles_1.styled)('p')({
    textAlign: 'center',
    width: '100%',
}, function (_a) {
    var style = _a.style;
    return ({
        borderTop: "1px solid ".concat(style.errorColor),
        color: style.errorColor,
        height: (0, styles_1.px)(style.h),
        lineHeight: (0, styles_1.px)(style.h),
    });
}, 'ErrorRSWP');
function ErrorMessage(_a) {
    var children = _a.children, _b = _a.styles, errorColor = _b.errorColor, height = _b.height;
    return (0, jsx_runtime_1.jsx)(Wrapper, __assign({ style: { errorColor: errorColor, h: height } }, { children: children }));
}
exports.default = ErrorMessage;
